### DISCUSSION TITLE
Skia Bindings build taking 18 minutes or more

### DISCUSSION THREAD

**Discussion by Xyndra** [rust-skia/rust-skia#1057] [General]
The C(++) compilation using clang-cl pushes my CPU to 100% and takes around 18 minutes
![image](https://github.com/user-attachments/assets/43ab5b45-5c1b-47a3-bf3e-26771aa41d0a)
See neofetch for my system
![image](https://github.com/user-attachments/assets/3016164a-a84a-4351-98dd-bee9dc026e1c)
It would be extremely nice if something like pre-compiled libraries would be added since this seems to happen for every recompile and it's needless to say that this is an unacceptably long time

**Comment by tronical** [comment]
> The C(++) compilation using clang-cl pushes my CPU to 100% and takes around 18 minutes ![image](https://private-user-images.githubusercontent.com/71836523/382518304-43ab5b45-5c1b-47a3-bf3e-26771aa41d0a.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzA2MDY4MTQsIm5iZiI6MTczMDYwNjUxNCwicGF0aCI6Ii83MTgzNjUyMy8zODI1MTgzMDQtNDNhYjViNDUtNWMxYi00N2EzLWJmM2UtMjY3NzFhYTQxZDBhLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDExMDMlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQxMTAzVDA0MDE1NFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWQ3OTllMWQyOTFhM2RlZjUyYzFlNzFkYzE5MTQyZDA0MDViODJlNzk2Y2I1ZjVjYjBlMzc2NWIxNjc0YTZhYjYmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.XnYo98WZF15fxIZgYU0qfsqmu7NH6FMVI5jz_m1stMM) See neofetch for my system ![image](https://private-user-images.githubusercontent.com/71836523/382518316-3016164a-a84a-4351-98dd-bee9dc026e1c.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzA2MDY4MTQsIm5iZiI6MTczMDYwNjUxNCwicGF0aCI6Ii83MTgzNjUyMy8zODI1MTgzMTYtMzAxNjE2NGEtYTg0YS00MzUxLTk4ZGQtYmVlOWRjMDI2ZTFjLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDExMDMlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQxMTAzVDA0MDE1NFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTQxNDBiY2JmMWRmMWViNzgxODEwZWRkZjhmZjMzZDNlY2Q4OTA1MmIzYmY1N2U1YjAzNTk1ZDBmMTFkZWE4YzEmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.vqoJkvpZmdGqqE5R5liy0RW8LszKNLUWrisyrHPAVgY) It would be extremely nice if something like pre-compiled libraries would be added

There are binaries for certain configurations and combinations of crate features:

https://github.com/rust-skia/rust-skia?tab=readme-ov-file#platform-support-build-targets-and-prebuilt-binaries

It would seem that the features that you’re activating ends up not being backed by binaries.

> since this seems to happen for every recompile 

I can’t reproduce that. For me, once built, a subsequent `cargo build` doesn‘t recompile skia. How does your reconciliation look like?

> and it's needless to say that this is an unacceptably long time

I’m not sure that there’s anything that can be done here. The build is parallelized with ninja, and it is a large code base. What would be an acceptable build time? How fast does it build for you if you build Skia upstream yourself?

**Comment by Xyndra** [comment]
I am using
```toml
skia-safe = {version = "0.78.2", features = ["all-windows"]}
```
on win11

For me, an acceptable time would be everything under 1:30 minutes. If it were to compile once and then cache reliably, I would understand, if the READMEs had a warning the first build takes long. But it is taking ages, and building the upstream has the same problem, and it runs on every cargo sync or cargo build. The binaries for certain configurations also don't seem to be hitting even though I am using `x86_64-pc-windows-msvc` with win11.

**Comment by Xyndra** [comment]
Note that after doing
```toml
[profile.dev]
incremental = true

[profile.release]
incremental = true
```
it does not recompile anymore, however this should be automatic
